sdks/python/apache_beam/dataframe/transforms.py

```diff
@@ -0,0 +1,246 @@
+              lambda _, inputs: {k: pd.concat(vs)
+                                 for k, vs in inputs.items()})
+        else:
+          # Already partitioned, or no partitioning needed.
+          (k, pcoll), = pcolls.items()
+          partitioned_pcoll = pcoll | beam.Map(lambda df: {k: df})
+
+        # Actually evaluate the expressions.
+        def evaluate(partition, stage=self.stage):
+          session = expressions.Session(
+              {expr: partition[expr._id]
+               for expr in stage.inputs})
+          for expr in stage.outputs:
+            yield beam.pvalue.TaggedOutput(expr._id, expr.evaluate_at(session))
+
+        return partitioned_pcoll | beam.FlatMap(evaluate).with_outputs()
+
+    class Stage(object):
+      """Used to build up a set of operations that can be fused together.
+      """
+      def __init__(self, inputs, is_grouping):
+        self.inputs = set(inputs)
+        self.is_grouping = is_grouping or len(self.inputs) > 1
+        self.ops = []
+        self.outputs = set()
+
+    # First define some helper functions.
+    def output_is_partitioned_by_index(expr, stage):
+      if expr in stage.inputs:
+        return stage.is_grouping
+      elif expr.preserves_partition_by_index():
+        if expr.requires_partition_by_index():
+          return True
+        else:
+          return all(
+              output_is_partitioned_by_index(arg, stage) for arg in expr.args())
+      else:
+        return False
+
+    def partition_by_index(df, levels=None, parts=10):
+      if levels is None:
+        levels = list(range(df.index.nlevels))
+      elif isinstance(levels, (int, str)):
+        levels = [levels]
+      hashes = sum(
+          pd.util.hash_array(df.index.get_level_values(level))
+          for level in levels)
+      for k in range(parts):
+        yield k, df[hashes % parts == k]
+
+    def common_stages(stage_lists):
+      if stage_lists:
+        for stage in stage_lists[0]:
+          if all(stage in other for other in stage_lists[1:]):
+            yield stage
+
+    @memoize
+    def expr_to_stages(expr):
+      assert expr not in inputs
+      # First attempt to compute this expression as part of an existing stage,
+      # if possible.
+      for stage in common_stages([expr_to_stages(arg) for arg in expr.args()
+                                  if arg not in inputs]):
+        if (not expr.requires_partition_by_index() or
+            all(output_is_partitioned_by_index(arg, stage)
+                for arg in expr.args())):
+          break
+      else:
+        # Otherwise, compute this expression as part of a new stage.
+        stage = Stage(expr.args(), expr.requires_partition_by_index())
+        for arg in expr.args():
+          if arg not in inputs:
+            expr_to_stages(arg).append(stage)
+            expr_to_stage(arg).outputs.add(arg)
+      stage.ops.append(expr)
+      # This is a list as given expression may be available in many stages.
```

Comment:
Is this for a future optimization? Unless I'm missing something, this function never returns more than a one-element list

Reply:
This will be useful for future optimizations, but is still useful now (it can also accumulate more than one value due to the `expr_to_stages(arg).append(stage)` line above).